### PR TITLE
Add Ingress NGINX v1.2.0 blogpost

### PR DESCRIPTION
**Pull request by rikatz**
Ingress NGINX v1.2.0 will come with some security improvements.

Due to this, the maintainers decided to write a blog post about it before the release (to be done on monday, 18 of April) so users can know a bit more about security improvements and what are the next steps.

### CODE REVIEW

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+## Meet Ingress NGINX v1.2.0 and the chrooted NGINX process
+
+One of the main problems is that we run our proxy server (NGINX) together with the controller (the component that has access to Kubernetes API Server and created the nginx.conf file). 
+
+So, NGINX does have the same access to the filesystem of the controller (and Kubernetes service account token, and other configurations from the container). While splitting those components is our end goal, we needed a fast response, that lead to the chroot idea.
+
+Let's take a look into what an Ingress Controller container looked like before this change:
+
+![Ingress NGINX pre chroot](ingress-pre-chroot.png)]
```
Comment (strongjz) [inline content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:26]: need to remove the ]

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+As we can see, the same container (not the Pod, the container!) that provides HTTP Proxy is the one that watches Ingress objects and writes the Container Volume
+
+Now, meet the new architecture:
+
+![Ingress NGINX post chroot](ingress-post-chroot.png)]
```
Comment (strongjz) [inline content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:32]: same here, ]

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+The one in left side is not chrooted. So NGINX has full access to the filesystem. The one in right side is chrooted, so a new filesystem with only the required files to make NGINX work is created.
+
+## What about other security improvements in this release?
+
+We know that the chroot improvement solves some portion of the risk, but still, someone can try to inject commands to read, for example, nginx.conf file and extract sensitive information.
+
+So, another change in this release (this is opt-out!) is the deep inspector. We know some directives and regexes may be dangerous to NGINX, so the deep inspector will check all fields from an Ingress object (during its reconciliation, and also with admission webhook!) to verify if some field contains these dangerous directives.
+
+We already do this for annotations, and the plan/goal is to move this validation to inside deep inspection in future releases.
```
Comment (sftim) [inline content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:110]:
```suggestion
The ingress controller already does this for annotations, and our hope is to move this existing validation to happen inside
deep inspection as part of a future release.

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+You can take a look into the existing rules in <https://github.com/kubernetes/ingress-nginx/blob/main/internal/ingress/inspector/rules.go>
```
Comment (sftim) [inline content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:112]:
```suggestion
You can take a look into the existing rules in [https://github.com/kubernetes/ingress-nginx/blob/main/internal/ingress/inspector/rules.go](https://github.com/kubernetes/ingress-nginx/blob/main/internal/ingress/inspector/rules.go).
```

Could this link instead be to https://github.com/kubernetes/ingress-nginx/blob/v1.20/internal/ingress/inspector/rules.go ?

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+Due to the nature of inspecting and matching all strings in all objects, this new feature may consume a bit more CPU, and can be disabled with the flag --deep-inspect=false
```
Comment (sftim) [inline content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:114]: Is this right:
```suggestion
Due to the nature of inspecting and matching all strings within relevant Ingress objects, this new feature may consume a bit more CPU. You can disable it by running the ingress controller with the command line argument `--deep-inspect=false`.
```

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+## What's next?
+
+This is not our final goal. Our final goal is to split the control plane and the data plane processes. In fact, doing so will help us also achieve the GatewayAPI implementation, as we may have a different controller as soon as it "knows" what to provide to the Data Plane (we need some help here!!)
+
+Some other projects in Kubernetes already take this approach (like [KPNG, the new Kube Proxy](​​https://github.com/kubernetes-sigs/kpng) and we plan to align with them and get the same experience for Ingress NGINX
```
Comment (sftim) [inline content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:120]:
```suggestion
Some other projects in Kubernetes already take this approach
(like [KPNG](​​https://github.com/kubernetes-sigs/kpng), the proposed replacement for `kube-proxy`),
and we plan to align with them and get the same experience for Ingress-NGINX.
```

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+## References
+
+References
+
+If you want to take a look into how chrooting was done in Ingress NGINX, take a look into <https://github.com/kubernetes/ingress-nginx/pull/8337>
+The release v1.2.0 containing all the changes can be found at <https://github.com/kubernetes/ingress-nginx/releases/tag/controller-v1.2.0>
```
Comment (sftim) [inline content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:127]: Please use `[]()` style Markdown hyperlinks here.

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+## What's next?
+
+This is not our final goal. Our final goal is to split the control plane and the data plane processes. In fact, doing so will help us also achieve the GatewayAPI implementation, as we may have a different controller as soon as it "knows" what to provide to the Data Plane (we need some help here!!)
+
+Some other projects in Kubernetes already take this approach (like [KPNG, the new Kube Proxy](​​https://github.com/kubernetes-sigs/kpng) and we plan to align with them and get the same experience for Ingress NGINX
+
+## References
+
+References
```
Comment (sftim) [inline content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:124]: Skip this line

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+Due to the nature of inspecting and matching all strings in all objects, this new feature may consume a bit more CPU, and can be disabled with the flag --deep-inspect=false
+
+## What's next?
+
+This is not our final goal. Our final goal is to split the control plane and the data plane processes. In fact, doing so will help us also achieve the GatewayAPI implementation, as we may have a different controller as soon as it "knows" what to provide to the Data Plane (we need some help here!!)
+
+Some other projects in Kubernetes already take this approach (like [KPNG, the new Kube Proxy](​​https://github.com/kubernetes-sigs/kpng) and we plan to align with them and get the same experience for Ingress NGINX
+
+## References
```
Comment (sftim) [inline content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:122]:
```suggestion
## Further reading
``

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+We already do this for annotations, and the plan/goal is to move this validation to inside deep inspection in future releases.
+
+You can take a look into the existing rules in <https://github.com/kubernetes/ingress-nginx/blob/main/internal/ingress/inspector/rules.go>
+
+Due to the nature of inspecting and matching all strings in all objects, this new feature may consume a bit more CPU, and can be disabled with the flag --deep-inspect=false
+
+## What's next?
+
+This is not our final goal. Our final goal is to split the control plane and the data plane processes. In fact, doing so will help us also achieve the GatewayAPI implementation, as we may have a different controller as soon as it "knows" what to provide to the Data Plane (we need some help here!!)
```
Comment (sftim) [inline content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:118]:
```suggestion
This is not our final goal. Our final goal is to split the control plane and the data plane processes.
In fact, doing so will help us also achieve a [Gateway](https://gateway-api.sigs.k8s.io/) API implementation,
as we may have a different controller as soon as it "knows" what to provide to the data plane
(we need some help here!!)
```

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+![Ingress NGINX pre chroot](ingress-pre-chroot.png)
+
+As we can see, the same container (not the Pod, the container!) that provides HTTP Proxy is the one that watches Ingress objects and writes the Container Volume
+
+Now, meet the new architecture:
+
+![Ingress NGINX post chroot](ingress-post-chroot.png)
+
+What does all of this mean? A good summary is: that we are running a new container inside our own container.
```
Comment (sftim) [inline content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:34]:
```suggestion
What does all of this mean? A basic summary is: that we are isolating the NGINX webserver as a container inside the
controller container.
```

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+slug: ingress-nginx-security-release
+---
+
+**Authors:** Ricardo Katz (VMware), James Strong (Chainguard)
+
+Ingress may be one of the most targeted components of Kubernetes. In fact, it usually is an HTTP Proxy, exposed to the Internet, containing multiple websites, and with some privileged access to Kubernetes API (mainly secret access for TLS certificates).
+
+While it is a risky component in your architecture, it is still required to properly expose your services.
+
+Ingress NGINX has been part of security assessments that figured out we have a big problem: we don't do all proper sanitization before turning the configuration into an nginx.conf file, which may lead to users being able to extract information.
```
Comment (sftim) [inline content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:14]:
```suggestion
Ingress-NGINX has been part of security assessments that figured out we have a big problem: we don't
do all proper sanitization before turning the configuration into an `nginx.conf` file, which may lead to information
disclosure risks.
```

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+
+While we understand this risk and the real need to fix this, it's not an easy process to do, so we took another approach to reduce (but not remove!) this risk in the next release.
```
Comment (sftim) [inline content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:16]:
```suggestion
While we understand this risk and the real need to fix this, it's not an easy process to do, so we took another approach to reduce (but not remove!) this risk in the current (v1.2.0) release.
```

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+title: 'Increasing the security bar in Ingress-NGINX'
+date: 2022-04-15
+slug: ingress-nginx-security-release
+---
+
+**Authors:** Ricardo Katz (VMware), James Strong (Chainguard)
+
+Ingress may be one of the most targeted components of Kubernetes. In fact, it usually is an HTTP Proxy, exposed to the Internet, containing multiple websites, and with some privileged access to Kubernetes API (mainly secret access for TLS certificates).
+
+While it is a risky component in your architecture, it is still required to properly expose your services.
```
Comment (sftim) [inline content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:12]: You don't need an Ingress to expose services (you could use Service or Gateway, for example). Maybe reword “required” to “the most popular way”.

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+
+While we understand this risk and the real need to fix this, it's not an easy process to do, so we took another approach to reduce (but not remove!) this risk in the next release.
+
+## Meet Ingress NGINX v1.2.0 and the chrooted NGINX process
+
+One of the main problems is that we run our proxy server (NGINX) together with the controller (the component that has access to Kubernetes API Server and created the nginx.conf file). 
+
+So, NGINX does have the same access to the filesystem of the controller (and Kubernetes service account token, and other configurations from the container). While splitting those components is our end goal, we needed a fast response, that lead to the chroot idea.
+
+Let's take a look into what an Ingress Controller container looked like before this change:
```
Comment (sftim) [inline content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:24]:
```suggestion
Let's take a look into what an NGINX-Ingress container looked like before this change:
```

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md [hunk 16]
```diff
@@ -0,0 +1,127 @@
+---
+layout: blog
+title: 'Increasing the security bar in Ingress-NGINX'
+date: 2022-04-15
+slug: ingress-nginx-security-release
```
Comment (sftim) [inline content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:5]:
```suggestion
slug: ingress-nginx-1-2-0
```

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md [hunk 17]
```diff
@@ -0,0 +1,127 @@
+---
+layout: blog
+title: 'Increasing the security bar in Ingress-NGINX'
```
Comment (sftim) [inline content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:3]:
```suggestion
title: 'Increasing the security bar in Ingress-NGINX v1.2.0
```

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+date: 2022-04-15
```
Comment (sftim) [inline content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:4]:
```suggestion
date: 2022-04-18
```
This is the release day, right?

**Review by sftim — COMMENTED** [review]
Thanks. Here's a bunch of feedback.

If you have time, replace these diagrams with SVG images that use Kubernetes colors (black, white, Kubernetes blue, gray). Convert the text to paths, too - that helps avoid rendering problems.

https://www.k8s.dev/blog/2021/12/01/improve-your-documentation-with-mermaid.js-diagrams/ is _one way_ to help with that work, but you definitely are also fine taking the existing vector source, tweaking it, and saving as SVG.

**Comment by sftim** [comment]
If this isn't done for the release day itself, it's OK to reword this slightly so that it's framed as a post-release article.

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md (hunk 15 above)
Comment (rikatz) [reply content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:28]: text here should be Ingress-Nginx actually. I will fix it manually on a next commit

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md (hunk 11 above)
Comment (rikatz) [reply content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:38]: s/webserver/proxy (or even isolating the NGINX Service). 

Will fix on another commit

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+
+![Ingress NGINX post chroot](ingress-post-chroot.png)
+
+What does all of this mean? A good summary is: that we are running a new container inside our own container.
+
+While this is not actually true, to understand what was done here, it's good to understand how Linux Containers (and kernel namespaces) work, and we recommend you read the article <https://www.nginx.com/blog/what-are-namespaces-cgroups-how-do-they-work/>.
+
+## Skip the talk, what do I need to use this new approach?
+
+While this increases the security, we made this feature an opt-in in this release so users may have time to make the right adjustments in their environments. This feature is only available from release v1.2.0
```
[the remark this answers is not in the document]
Comment (rikatz) [reply content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:49]: same here, should be Ingress-Nginx

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md (hunk 3 above)
Comment (rikatz) [reply content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:119]: I've changed the word hope to goal. Hope seems to much dramatic (in the sense of "we hope someday this will happen" instead of "we plan this to real happen")

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md (hunk 4 above)
Comment (rikatz) [reply content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:121]: yes and no. The problem in this case is, we may update this rules between versions. If someone come to the blog and check this and thinks "yey, I'm on v1.3.0, let me see how the rules are" they may end up seeing old rules.

Thoughts?

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md (hunk 5 above)
Comment (rikatz) [reply content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:123]: so, today the deep inspector does only Ingress objects but it's prepared for service objects (as an example, one may want to block localhost services for CVE reasons).

We may improve this tho, with your review but not discarding that other objects may be inspected

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md (hunk 15 above)
Comment (rikatz) [reply content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:30]: Nevermind, I could edit your comment and fix here

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md (hunk 11 above)
Comment (rikatz) [reply content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:40]: same, edited here

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md (hunk 5 above)
Comment (rikatz) [reply content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:137]: actually right now it does make sense. Let me commit here

File: content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+
+Let's take a look into what an Ingress Controller container looked like before this change:
+
+![Ingress NGINX pre chroot](ingress-pre-chroot.png)
+
+As we can see, the same container (not the Pod, the container!) that provides HTTP Proxy is the one that watches Ingress objects and writes the Container Volume
+
+Now, meet the new architecture:
+
+![Ingress NGINX post chroot](ingress-post-chroot.png)
```
[the remark this answers is not in the document]
Comment (rikatz) [reply content/en/blog/_posts/2022-04-15-Increasing-the-security-bar-in-Ingress-NGINX/index.md:38]: was going to edit here, but Mac have some hard time with SVG. If we can leave the diagrams fixing (even this, or turning into a mermaid one) for post publication, would appreciate that

**Comment by sftim** [comment]
How about the 28th as a pencilled-in publication date?

**Comment by rikatz** [comment]
@sftim Changed the publishing date to Apr/28 here, PTAL :)

Thank you!

**Comment by rikatz** [comment]
@sftim do we think we can release this today? :D ty

**Comment by sftim** [comment]
_dons blog editorial team hat_
/lgtm
/approve

**Comment by rikatz** [comment]
Thank you for being so responsive :D
